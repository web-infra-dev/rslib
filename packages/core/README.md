<picture>
  <img alt="Rslib Banner" src="https://assets.rspack.dev/rslib/rslib-banner.png">
</picture>

# Rslib

Rslib is a library build tool powered by [Rsbuild](https://rsbuild.dev). It allows library developers to leverage the knowledge and ecosystem of webpack and Rspack.

## Documentation

https://lib.rsbuild.dev/

## Contributing

Please read the [Contributing Guide](https://github.com/web-infra-dev/rslib/blob/main/CONTRIBUTING.md).

## License

Rslib is [MIT licensed](https://github.com/web-infra-dev/rslib/blob/main/LICENSE).
