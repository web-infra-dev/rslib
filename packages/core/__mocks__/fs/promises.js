import { promises } from 'memfs';
export default promises;
