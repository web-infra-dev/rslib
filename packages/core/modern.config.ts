import { defineConfig, moduleTools } from '@modern-js/module-tools';

export default defineConfig({
  plugins: [moduleTools()],
  buildConfig: [
    {
      input: ['src'],
      buildType: 'bundleless',
      format: 'cjs',
      autoExtension: true,
      target: 'es2020',
    },
  ],
});
