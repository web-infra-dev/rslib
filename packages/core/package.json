{
  "name": "@rslib/core",
  "version": "0.0.10",
  "description": "The Rspack-based library build tool.",
  "homepage": "https://lib.rsbuild.dev",
  "bugs": {
    "url": "https://github.com/web-infra-dev/rslib/issues"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/web-infra-dev/rslib",
    "directory": "packages/core"
  },
  "license": "MIT",
  "type": "module",
  "exports": {
    ".": {
      "types": "./dist-types/index.d.ts",
      "default": "./dist/index.js"
    },
    "./package.json": "./package.json"
  },
  "main": "./dist/index.js",
  "types": "./dist-types/index.d.ts",
  "bin": {
    "rslib": "./bin/rslib.js"
  },
  "files": [
    "bin",
    "dist",
    "dist-types",
    "compiled"
  ],
  "scripts": {
    "build": "rslib build",
    "dev": "rslib build --watch",
    "prebundle": "prebundle"
  },
  "dependencies": {
    "@rsbuild/core": "1.0.12",
    "rsbuild-plugin-dts": "workspace:*"
  },
  "devDependencies": {
    "@rslib/tsconfig": "workspace:*",
    "@rspack/core": "1.0.8",
    "@types/fs-extra": "^11.0.4",
    "commander": "^12.1.0",
    "fast-glob": "^3.3.2",
    "fs-extra": "^11.2.0",
    "memfs": "^4.13.0",
    "picocolors": "1.1.0",
    "prebundle": "1.2.2",
    "rslib": "npm:@rslib/core@0.0.10",
    "rslog": "^1.2.3",
    "tsconfck": "3.1.4",
    "typescript": "^5.6.3"
  },
  "peerDependencies": {
    "@microsoft/api-extractor": "^7",
    "typescript": "^5"
  },
  "peerDependenciesMeta": {
    "@microsoft/api-extractor": {
      "optional": true
    },
    "typescript": {
      "optional": true
    }
  },
  "engines": {
    "node": ">=16.0.0"
  },
  "publishConfig": {
    "access": "public",
    "provenance": true,
    "registry": "https://registry.npmjs.org/"
  }
}
