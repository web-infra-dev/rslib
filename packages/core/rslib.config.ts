import { defineConfig } from 'rslib';

export default defineConfig({
  lib: [
    {
      format: 'esm',
      syntax: ['node 16'],
      dts: {
        bundle: false,
        distPath: './dist-types',
      },
    },
  ],
  source: {
    entry: {
      index: './src/index.ts',
      libCssExtractLoader: './src/css/libCssExtractLoader.ts',
    },
    define: {
      RSLIB_VERSION: JSON.stringify(require('./package.json').version),
    },
  },
  output: {
    target: 'node',
    externals: {
      picocolors: '../compiled/picocolors/index.js',
      'fast-glob': '../compiled/fast-glob/index.js',
      commander: '../compiled/commander/index.js',
      rslog: '../compiled/rslog/index.js',
    },
  },
});
