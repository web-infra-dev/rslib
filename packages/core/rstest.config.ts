import { defineConfig } from '@rstest/core';
import { shared } from '../../rstest.workspace';

export default defineConfig({
  ...shared,
  name: 'unit',
});
