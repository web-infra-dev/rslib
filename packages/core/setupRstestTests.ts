import path from 'node:path';
import process from 'node:process';
import { beforeEach, expect } from '@rstest/core';
import { createSnapshotSerializer } from 'path-serializer';

beforeEach(() => {
  // since our NODE_ENV injection logic is via cli, we need to
  // delete "test" NODE_ENV to avoid affecting the default build config
  delete process.env.NODE_ENV;
});

function removeTrailingSlash(inputPath: string): string {
  return path.normalize(inputPath).replace(/[\\/]+$/, '');
}
expect.addSnapshotSerializer(
  createSnapshotSerializer({
    root: removeTrailingSlash(path.join(__dirname, '../../')),
    workspace: __dirname,
    features: {
      replaceHomeDir: false,
      replaceWorkspace: true,
      replaceRoot: true,
      escapeDoubleQuotes: false,
      transformCLR: false,
    },
  }),
);

if (process.env.ECO_CI) {
  expect.extend({
    toMatchSnapshot: () => ({
      pass: true,
      message: () => 'Snapshot always passes',
    }),
    toMatchInlineSnapshot: () => ({
      pass: true,
      message: () => 'Snapshot always passes',
    }),
  });
}
