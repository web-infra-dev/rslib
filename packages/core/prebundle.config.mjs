// @ts-check
import { join } from 'node:path';
import fs from 'fs-extra';

/**
 * Tip: please add the prebundled packages to `tsconfig.json#paths`.
 */

/** @type {import('prebundle').Config} */
export default {
  externals: {
    typescript: 'typescript',
  },
  dependencies: [
    'commander',
    'fast-glob',
    {
      name: 'rslog',
      afterBundle(task) {
        // use the cjs bundle of rslog
        fs.copyFileSync(
          join(task.depPath, 'dist/index.cjs'),
          join(task.distPath, 'index.js'),
        );
      },
    },
    {
      name: 'picocolors',
      beforeBundle({ depPath }) {
        const typesFile = join(depPath, 'types.ts');
        // Fix type bundle
        if (fs.existsSync(typesFile)) {
          fs.renameSync(typesFile, join(depPath, 'types.d.ts'));
        }
      },
    },
  ],
};
