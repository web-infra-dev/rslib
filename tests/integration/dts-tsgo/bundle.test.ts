import { spawnSync } from 'node:child_process';
import { existsSync } from 'node:fs';
import { join } from 'node:path';
import { describe, expect, test } from '@rstest/core';
import { buildAndGetResults, createTempFiles, queryContent } from 'test-helper';

describe.skipIf(process.version.startsWith('v18'))(
  'dts with tsgo when bundle: true',
  () => {
    test('basic', async () => {
      const fixturePath = join(__dirname, 'bundle', 'basic');
      const { files, entries } = await buildAndGetResults({
        fixturePath,
        type: 'dts',
      });

      expect(files.esm).toMatchInlineSnapshot(
        `
      [
        "<ROOT>/tests/integration/dts-tsgo/bundle/basic/dist/esm/index.d.ts",
      ]
    `,
      );

      expect(files.cjs).toMatchInlineSnapshot(
        `
      [
        "<ROOT>/tests/integration/dts-tsgo/bundle/basic/dist/cjs/index.d.ts",
      ]
    `,
      );

      expect(entries).toMatchSnapshot();
    });

    test('distPath', async () => {
      const fixturePath = join(__dirname, 'bundle', 'dist-path');
      const { files } = await buildAndGetResults({
        fixturePath,
        type: 'dts',
      });

      expect(files.esm).toMatchInlineSnapshot(
        `
      [
        "<ROOT>/tests/integration/dts-tsgo/bundle/dist-path/dist/custom/index.d.ts",
      ]
    `,
      );
    });

    test('abortOnError: false', async () => {
      const fixturePath = join(__dirname, 'bundle', 'abort-on-error');

      const result = spawnSync('npx', ['rslib', 'build'], {
        cwd: fixturePath,
        // do not show output in test console
        stdio: 'ignore',
        shell: true,
      });

      expect(result.status).toBe(0);
    });

    test('autoExtension: true', async () => {
      const fixturePath = join(__dirname, 'bundle', 'auto-extension');
      const { files } = await buildAndGetResults({
        fixturePath,
        type: 'dts',
      });

      expect(files.cjs).toMatchInlineSnapshot(
        `
      [
        "<ROOT>/tests/integration/dts-tsgo/bundle/auto-extension/dist/cjs/index.d.cts",
      ]
    `,
      );
    });

    test('bundleName -- set source.entry', async () => {
      const fixturePath = join(__dirname, 'bundle', 'bundle-name');
      const { files } = await buildAndGetResults({
        fixturePath,
        type: 'dts',
      });

      expect(files.esm).toMatchInlineSnapshot(
        `
      [
        "<ROOT>/tests/integration/dts-tsgo/bundle/bundle-name/dist/esm/bundleName.d.ts",
      ]
    `,
      );
    });

    test('entry is an absolute path', async () => {
      const fixturePath = join(__dirname, 'bundle', 'absolute-entry');
      const { files } = await buildAndGetResults({
        fixturePath,
        type: 'dts',
      });

      expect(files.esm).toMatchInlineSnapshot(
        `
      [
        "<ROOT>/tests/integration/dts-tsgo/bundle/absolute-entry/dist/esm/index.d.ts",
      ]
    `,
      );
    });

    test('rootdir calculation should ignore declaration files', async () => {
      const fixturePath = join(__dirname, 'bundle', 'rootdir');
      const { files, entries } = await buildAndGetResults({
        fixturePath,
        type: 'dts',
      });

      expect(files.esm).toMatchInlineSnapshot(`
      [
        "<ROOT>/tests/integration/dts-tsgo/bundle/rootdir/dist/esm/index.d.ts",
      ]
    `);

      expect(files.cjs).toMatchInlineSnapshot(`
      [
        "<ROOT>/tests/integration/dts-tsgo/bundle/rootdir/dist/cjs/index.d.ts",
      ]
    `);

      expect(entries).toMatchSnapshot();
    });

    test('should clean dts dist files and .rslib folder', async () => {
      const fixturePath = join(__dirname, 'bundle', 'clean');

      const checkFiles = await createTempFiles(fixturePath, true);

      const { files } = await buildAndGetResults({ fixturePath, type: 'dts' });

      for (const file of checkFiles) {
        expect(existsSync(file)).toBe(false);
      }

      expect(files.esm).toMatchInlineSnapshot(`
      [
        "<ROOT>/tests/integration/dts-tsgo/bundle/clean/dist-types/esm/index.d.ts",
      ]
    `);

      expect(files.cjs).toMatchInlineSnapshot(`
      [
        "<ROOT>/tests/integration/dts-tsgo/bundle/clean/dist-types/cjs/index.d.ts",
      ]
    `);
    });

    test('multiple entries', async () => {
      const fixturePath = join(__dirname, 'bundle', 'multiple-entries');
      const { files, contents } = await buildAndGetResults({
        fixturePath,
        type: 'dts',
      });

      expect(files.esm).toMatchInlineSnapshot(`
      [
        "<ROOT>/tests/integration/dts-tsgo/bundle/multiple-entries/dist/esm/index.d.ts",
        "<ROOT>/tests/integration/dts-tsgo/bundle/multiple-entries/dist/esm/sum.d.ts",
      ]
    `);

      expect(files.cjs).toMatchInlineSnapshot(`
      [
        "<ROOT>/tests/integration/dts-tsgo/bundle/multiple-entries/dist/cjs/index.d.ts",
        "<ROOT>/tests/integration/dts-tsgo/bundle/multiple-entries/dist/cjs/sum.d.ts",
      ]
    `);

      const { content: indexEsm } = queryContent(contents.esm, 'index.d.ts', {
        basename: true,
      });
      const { content: indexCjs } = queryContent(contents.cjs, 'index.d.ts', {
        basename: true,
      });
      const { content: sumEsm } = queryContent(contents.esm, 'sum.d.ts', {
        basename: true,
      });
      const { content: sumCjs } = queryContent(contents.cjs, 'sum.d.ts', {
        basename: true,
      });

      expect([indexEsm, indexCjs, sumEsm, sumCjs]).toMatchSnapshot();
    });

    test('override with bundledPackages', async () => {
      const fixturePath = join(__dirname, 'bundle', 'bundled-packages');
      const { entries } = await buildAndGetResults({
        fixturePath,
        type: 'dts',
      });

      // default
      expect(entries.esm0).toContain(`import { Action } from 'redux';`);

      // override empty array
      expect(entries.esm1).toMatchInlineSnapshot(`
      "
      export * from "@reduxjs/toolkit";

      export { }
      "
    `);

      // override with bundledPackages
      expect(entries.esm2).not.toContain(`import { Action } from 'redux';`);
    });
  },
);
