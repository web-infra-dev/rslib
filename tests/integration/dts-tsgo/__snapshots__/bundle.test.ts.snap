// Rstest Snapshot v1

exports[`dts with tsgo when bundle: true > basic 3`] = `
{
  "cjs": "export declare const num1 = 1;

export declare const num2 = 2;

export declare const num3 = 3;

export declare const numSum: number;

export declare const str1 = "str1";

export declare const str2 = "str2";

export declare const str3 = "str3";

export declare const strSum: string;

export { }
",
  "esm": "export declare const num1 = 1;

export declare const num2 = 2;

export declare const num3 = 3;

export declare const numSum: number;

export declare const str1 = "str1";

export declare const str2 = "str2";

export declare const str3 = "str3";

export declare const strSum: string;

export { }
",
}
`;

exports[`dts with tsgo when bundle: true > multiple entries 3`] = `
[
  "export declare const num1 = 1;

export declare const num2 = 2;

export declare const num3 = 3;

export declare const numSum: number;

export declare const str1 = "str1";

export declare const str2 = "str2";

export declare const str3 = "str3";

export declare const strSum: string;

export { }
",
  "export declare const num1 = 1;

export declare const num2 = 2;

export declare const num3 = 3;

export declare const numSum: number;

export declare const str1 = "str1";

export declare const str2 = "str2";

export declare const str3 = "str3";

export declare const strSum: string;

export { }
",
  "export declare const numSum: number;

export declare const strSum: string;

export { }
",
  "export declare const numSum: number;

export declare const strSum: string;

export { }
",
]
`;

exports[`dts with tsgo when bundle: true > rootdir calculation should ignore declaration files 3`] = `
{
  "cjs": "export declare const num1 = 1;

export declare const num2 = 2;

export declare const num3 = 3;

export declare const numSum: number;

export declare const str1 = "str1";

export declare const str2 = "str2";

export declare const str3 = "str3";

export declare const strSum: string;

export { }
",
  "esm": "export declare const num1 = 1;

export declare const num2 = 2;

export declare const num3 = 3;

export declare const numSum: number;

export declare const str1 = "str1";

export declare const str2 = "str2";

export declare const str3 = "str3";

export declare const strSum: string;

export { }
",
}
`;
